File: LicensedApps/FileLicenseControl/LicensedColorComboBox.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;

namespace FileLicenseControl
{
	/// <summary>
	/// Summary description for LicensedColorComboBox.
	/// </summary>
	[LicenseProvider(typeof(LicFileLicenseProvider))]
	public class LicensedColorComboBox : System.Windows.Forms.ComboBox
	{
		private License license = null;

		public LicensedColorComboBox()
		{
			license = LicenseManager.Validate(typeof(LicensedColorComboBox), this);
			FillItems();
			base.SelectedItem = base.Items[7]; // Black
			base.DrawItem += new DrawItemEventHandler(this.combo_DrawItem);
		}

		~LicensedColorComboBox() 
		{
			if (license != null) 
			{
				license.Dispose();
				license = null;
			}
		}

		public void FillItems()
		{
			// Fill the combo box with all of the colors in the Colors enumeration
			base.DisplayMember = "Color";

			base.Items.Add(Brushes.AliceBlue);
			base.Items.Add(Brushes.AntiqueWhite);
			base.Items.Add(Brushes.Aqua);
			base.Items.Add(Brushes.Aquamarine);
			base.Items.Add(Brushes.Azure);
			base.Items.Add(Brushes.Beige);
			base.Items.Add(Brushes.Bisque);
			base.Items.Add(Brushes.Black);
			base.Items.Add(Brushes.BlanchedAlmond);
			base.Items.Add(Brushes.Blue);
			base.Items.Add(Brushes.BlueViolet);
			base.Items.Add(Brushes.Brown);
			base.Items.Add(Brushes.BurlyWood);
			base.Items.Add(Brushes.CadetBlue);
			base.Items.Add(Brushes.Chartreuse);
			base.Items.Add(Brushes.Chocolate);
			base.Items.Add(Brushes.Coral);
			base.Items.Add(Brushes.CornflowerBlue);
			base.Items.Add(Brushes.Cornsilk);
			base.Items.Add(Brushes.Crimson);
			base.Items.Add(Brushes.Cyan);
			base.Items.Add(Brushes.DarkBlue);
			base.Items.Add(Brushes.DarkCyan);
			base.Items.Add(Brushes.DarkGoldenrod);
			base.Items.Add(Brushes.DarkGray);
			base.Items.Add(Brushes.DarkGreen);
			base.Items.Add(Brushes.DarkKhaki);
			base.Items.Add(Brushes.DarkMagenta);
			base.Items.Add(Brushes.DarkOliveGreen);
			base.Items.Add(Brushes.DarkOrange);
			base.Items.Add(Brushes.DarkOrchid);
			base.Items.Add(Brushes.DarkRed);
			base.Items.Add(Brushes.DarkSalmon);
			base.Items.Add(Brushes.DarkSeaGreen);
			base.Items.Add(Brushes.DarkSlateBlue);
			base.Items.Add(Brushes.DarkSlateGray);
			base.Items.Add(Brushes.DarkTurquoise);
			base.Items.Add(Brushes.DarkViolet);
			base.Items.Add(Brushes.DeepPink);
			base.Items.Add(Brushes.DeepSkyBlue);
			base.Items.Add(Brushes.DimGray);
			base.Items.Add(Brushes.DodgerBlue);
			base.Items.Add(Brushes.Firebrick);
			base.Items.Add(Brushes.FloralWhite);
			base.Items.Add(Brushes.ForestGreen);
			base.Items.Add(Brushes.Fuchsia);
			base.Items.Add(Brushes.Gainsboro);
			base.Items.Add(Brushes.GhostWhite);
			base.Items.Add(Brushes.Gold);
			base.Items.Add(Brushes.Goldenrod);
			base.Items.Add(Brushes.Gray);
			base.Items.Add(Brushes.Green);
			base.Items.Add(Brushes.GreenYellow);
			base.Items.Add(Brushes.Honeydew);
			base.Items.Add(Brushes.HotPink);
			base.Items.Add(Brushes.IndianRed);
			base.Items.Add(Brushes.Indigo);
			base.Items.Add(Brushes.Ivory);
			base.Items.Add(Brushes.Khaki);
			base.Items.Add(Brushes.Lavender);
			base.Items.Add(Brushes.LavenderBlush);
			base.Items.Add(Brushes.LawnGreen);
			base.Items.Add(Brushes.LemonChiffon);
			base.Items.Add(Brushes.LightBlue);
			base.Items.Add(Brushes.LightCoral);
			base.Items.Add(Brushes.LightCyan);
			base.Items.Add(Brushes.LightGoldenrodYellow);
			base.Items.Add(Brushes.LightGray);
			base.Items.Add(Brushes.LightGreen);
			base.Items.Add(Brushes.LightPink);
			base.Items.Add(Brushes.LightSalmon);
			base.Items.Add(Brushes.LightSeaGreen);
			base.Items.Add(Brushes.LightSkyBlue);
			base.Items.Add(Brushes.LightSlateGray);
			base.Items.Add(Brushes.LightSteelBlue);
			base.Items.Add(Brushes.LightYellow);
			base.Items.Add(Brushes.Lime);
			base.Items.Add(Brushes.LimeGreen);
			base.Items.Add(Brushes.Linen);
			base.Items.Add(Brushes.Magenta);
			base.Items.Add(Brushes.Maroon);
			base.Items.Add(Brushes.MediumAquamarine);
			base.Items.Add(Brushes.MediumBlue);
			base.Items.Add(Brushes.MediumOrchid);
			base.Items.Add(Brushes.MediumPurple);
			base.Items.Add(Brushes.MediumSeaGreen);
			base.Items.Add(Brushes.MediumSlateBlue);
			base.Items.Add(Brushes.MediumSpringGreen);
			base.Items.Add(Brushes.MediumTurquoise);
			base.Items.Add(Brushes.MediumVioletRed);
			base.Items.Add(Brushes.MidnightBlue);
			base.Items.Add(Brushes.MintCream);
			base.Items.Add(Brushes.MistyRose);
			base.Items.Add(Brushes.Moccasin);
			base.Items.Add(Brushes.NavajoWhite);
			base.Items.Add(Brushes.Navy);
			base.Items.Add(Brushes.OldLace);
			base.Items.Add(Brushes.Olive);
			base.Items.Add(Brushes.OliveDrab);
			base.Items.Add(Brushes.Orange);
			base.Items.Add(Brushes.OrangeRed);
			base.Items.Add(Brushes.Orchid);
			base.Items.Add(Brushes.PaleGoldenrod);
			base.Items.Add(Brushes.PaleGreen);
			base.Items.Add(Brushes.PaleTurquoise);
			base.Items.Add(Brushes.PaleVioletRed);
			base.Items.Add(Brushes.PapayaWhip);
			base.Items.Add(Brushes.PeachPuff);
			base.Items.Add(Brushes.Peru);
			base.Items.Add(Brushes.Pink);
			base.Items.Add(Brushes.Plum);
			base.Items.Add(Brushes.PowderBlue);
			base.Items.Add(Brushes.Purple);
			base.Items.Add(Brushes.Red);
			base.Items.Add(Brushes.RosyBrown);
			base.Items.Add(Brushes.RoyalBlue);
			base.Items.Add(Brushes.SaddleBrown);
			base.Items.Add(Brushes.Salmon);
			base.Items.Add(Brushes.SandyBrown);
			base.Items.Add(Brushes.SeaGreen);
			base.Items.Add(Brushes.SeaShell);
			base.Items.Add(Brushes.Sienna);
			base.Items.Add(Brushes.Silver);
			base.Items.Add(Brushes.SkyBlue);
			base.Items.Add(Brushes.SlateBlue);
			base.Items.Add(Brushes.SlateGray);
			base.Items.Add(Brushes.Snow);
			base.Items.Add(Brushes.SpringGreen);
			base.Items.Add(Brushes.SteelBlue);
			base.Items.Add(Brushes.Tan);
			base.Items.Add(Brushes.Teal);
			base.Items.Add(Brushes.Thistle);
			base.Items.Add(Brushes.Tomato);
			base.Items.Add(Brushes.Transparent);
			base.Items.Add(Brushes.Turquoise);
			base.Items.Add(Brushes.Violet);
			base.Items.Add(Brushes.Wheat);
			base.Items.Add(Brushes.White);
			base.Items.Add(Brushes.WhiteSmoke);
			base.Items.Add(Brushes.Yellow);
			base.Items.Add(Brushes.YellowGreen);
		}

		public Color Color
		{
			get { return ((SolidBrush)base.SelectedItem).Color; }
			set
			{
				string strInColor = value.ToString();
				foreach ( SolidBrush myColor in base.Items )
				{
					// Check the string
					if ( String.Compare(strInColor, myColor.Color.ToString()) == 0 )
					{
						// Found it
						base.SelectedItem = myColor;
						break;
					} // if
				} // foreach
			}
		}

		private void combo_DrawItem(object sender, System.Windows.Forms.DrawItemEventArgs die)
		{
			// If we have no index to draw, return.
			if ( die.Index == -1 )
			{
				return;
			} // if

			// The "sender" will be a combobox instance...if none, return.
			if ( sender == null )
			{
				return;
			} // if

			// Pull the controlling combobox object and prepare to render...
			ComboBox cmb = (ComboBox)sender;
			SolidBrush selectedBrush = (SolidBrush)cmb.Items[die.Index];
			Graphics g = die.Graphics;

			// If the item is selected, this draws the correct background color
			die.DrawBackground();
			die.DrawFocusRectangle();
			Rectangle rectPreviewBox = die.Bounds;

			// Draw the color's preview box
			rectPreviewBox.Offset(2, 2);
			rectPreviewBox.Width = 20;
			rectPreviewBox.Height -= 4;
			g.DrawRectangle(new Pen(die.ForeColor), rectPreviewBox);

			// Get the appropriate Brush object for the selected color
			// and fill the preview box.
			rectPreviewBox.Offset(1, 1);
			rectPreviewBox.Width -= 2;
			rectPreviewBox.Height -= 2;
			g.FillRectangle(selectedBrush, rectPreviewBox);

			// Draw the name of the color selected. Note we'll clean up the returned
			// type string. Normally it would come back:
			//
			// "Color [Black]"
			//
			// We'll clean it up to produce:
			//
			// "Black"
			g.DrawString(selectedBrush.Color.ToString().Substring(7,selectedBrush.Color.ToString().Length-8), Font, new SolidBrush(die.ForeColor), die.Bounds.X + 30, die.Bounds.Y + 1);
		}
	}
}

File: LicensedApps/VBFileLicenseControl/LicensedColorComboBox.vb
Imports System
Imports System.Windows.Forms
Imports System.Drawing
Imports System.ComponentModel

<LicenseProviderAttribute(GetType(LicFileLicenseProvider))> _
Public Class LicensedColorComboBox
    Inherits System.Windows.Forms.ComboBox

    Dim _license As License = Nothing

    Public Sub New()
        MyBase.New()
        _license = LicenseManager.Validate(GetType(LicensedColorComboBox), Me)

        FillItems()
        Me.SelectedItem = Items(7) ' Black
        AddHandler Me.DrawItem, New System.Windows.Forms.DrawItemEventHandler(AddressOf Me.DrawItems)
    End Sub

    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
        If Not _license Is Nothing Then
            _license.Dispose()
            _license = Nothing
        End If

        MyBase.Dispose(disposing)
    End Sub

    Protected Sub FillItems()
        Me.DisplayMember = "Color"

        Me.Items.Add(Brushes.AliceBlue)
        Me.Items.Add(Brushes.AntiqueWhite)
        Me.Items.Add(Brushes.Aqua)
        Me.Items.Add(Brushes.Aquamarine)
        Me.Items.Add(Brushes.Azure)
        Me.Items.Add(Brushes.Beige)
        Me.Items.Add(Brushes.Bisque)
        Me.Items.Add(Brushes.Black)
        Me.Items.Add(Brushes.BlanchedAlmond)
        Me.Items.Add(Brushes.Blue)
        Me.Items.Add(Brushes.BlueViolet)
        Me.Items.Add(Brushes.Brown)
        Me.Items.Add(Brushes.BurlyWood)
        Me.Items.Add(Brushes.CadetBlue)
        Me.Items.Add(Brushes.Chartreuse)
        Me.Items.Add(Brushes.Chocolate)
        Me.Items.Add(Brushes.Coral)
        Me.Items.Add(Brushes.CornflowerBlue)
        Me.Items.Add(Brushes.Cornsilk)
        Me.Items.Add(Brushes.Crimson)
        Me.Items.Add(Brushes.Cyan)
        Me.Items.Add(Brushes.DarkBlue)
        Me.Items.Add(Brushes.DarkCyan)
        Me.Items.Add(Brushes.DarkGoldenrod)
        Me.Items.Add(Brushes.DarkGray)
        Me.Items.Add(Brushes.DarkGreen)
        Me.Items.Add(Brushes.DarkKhaki)
        Me.Items.Add(Brushes.DarkMagenta)
        Me.Items.Add(Brushes.DarkOliveGreen)
        Me.Items.Add(Brushes.DarkOrange)
        Me.Items.Add(Brushes.DarkOrchid)
        Me.Items.Add(Brushes.DarkRed)
        Me.Items.Add(Brushes.DarkSalmon)
        Me.Items.Add(Brushes.DarkSeaGreen)
        Me.Items.Add(Brushes.DarkSlateBlue)
        Me.Items.Add(Brushes.DarkSlateGray)
        Me.Items.Add(Brushes.DarkTurquoise)
        Me.Items.Add(Brushes.DarkViolet)
        Me.Items.Add(Brushes.DeepPink)
        Me.Items.Add(Brushes.DeepSkyBlue)
        Me.Items.Add(Brushes.DimGray)
        Me.Items.Add(Brushes.DodgerBlue)
        Me.Items.Add(Brushes.Firebrick)
        Me.Items.Add(Brushes.FloralWhite)
        Me.Items.Add(Brushes.ForestGreen)
        Me.Items.Add(Brushes.Fuchsia)
        Me.Items.Add(Brushes.Gainsboro)
        Me.Items.Add(Brushes.GhostWhite)
        Me.Items.Add(Brushes.Gold)
        Me.Items.Add(Brushes.Goldenrod)
        Me.Items.Add(Brushes.Gray)
        Me.Items.Add(Brushes.Green)
        Me.Items.Add(Brushes.GreenYellow)
        Me.Items.Add(Brushes.Honeydew)
        Me.Items.Add(Brushes.HotPink)
        Me.Items.Add(Brushes.IndianRed)
        Me.Items.Add(Brushes.Indigo)
        Me.Items.Add(Brushes.Ivory)
        Me.Items.Add(Brushes.Khaki)
        Me.Items.Add(Brushes.Lavender)
        Me.Items.Add(Brushes.LavenderBlush)
        Me.Items.Add(Brushes.LawnGreen)
        Me.Items.Add(Brushes.LemonChiffon)
        Me.Items.Add(Brushes.LightBlue)
        Me.Items.Add(Brushes.LightCoral)
        Me.Items.Add(Brushes.LightCyan)
        Me.Items.Add(Brushes.LightGoldenrodYellow)
        Me.Items.Add(Brushes.LightGray)
        Me.Items.Add(Brushes.LightGreen)
        Me.Items.Add(Brushes.LightPink)
        Me.Items.Add(Brushes.LightSalmon)
        Me.Items.Add(Brushes.LightSeaGreen)
        Me.Items.Add(Brushes.LightSkyBlue)
        Me.Items.Add(Brushes.LightSlateGray)
        Me.Items.Add(Brushes.LightSteelBlue)
        Me.Items.Add(Brushes.LightYellow)
        Me.Items.Add(Brushes.Lime)
        Me.Items.Add(Brushes.LimeGreen)
        Me.Items.Add(Brushes.Linen)
        Me.Items.Add(Brushes.Magenta)
        Me.Items.Add(Brushes.Maroon)
        Me.Items.Add(Brushes.MediumAquamarine)
        Me.Items.Add(Brushes.MediumBlue)
        Me.Items.Add(Brushes.MediumOrchid)
        Me.Items.Add(Brushes.MediumPurple)
        Me.Items.Add(Brushes.MediumSeaGreen)
        Me.Items.Add(Brushes.MediumSlateBlue)
        Me.Items.Add(Brushes.MediumSpringGreen)
        Me.Items.Add(Brushes.MediumTurquoise)
        Me.Items.Add(Brushes.MediumVioletRed)
        Me.Items.Add(Brushes.MidnightBlue)
        Me.Items.Add(Brushes.MintCream)
        Me.Items.Add(Brushes.MistyRose)
        Me.Items.Add(Brushes.Moccasin)
        Me.Items.Add(Brushes.NavajoWhite)
        Me.Items.Add(Brushes.Navy)
        Me.Items.Add(Brushes.OldLace)
        Me.Items.Add(Brushes.Olive)
        Me.Items.Add(Brushes.OliveDrab)
        Me.Items.Add(Brushes.Orange)
        Me.Items.Add(Brushes.OrangeRed)
        Me.Items.Add(Brushes.Orchid)
        Me.Items.Add(Brushes.PaleGoldenrod)
        Me.Items.Add(Brushes.PaleGreen)
        Me.Items.Add(Brushes.PaleTurquoise)
        Me.Items.Add(Brushes.PaleVioletRed)
        Me.Items.Add(Brushes.PapayaWhip)
        Me.Items.Add(Brushes.PeachPuff)
        Me.Items.Add(Brushes.Peru)
        Me.Items.Add(Brushes.Pink)
        Me.Items.Add(Brushes.Plum)
        Me.Items.Add(Brushes.PowderBlue)
        Me.Items.Add(Brushes.Purple)
        Me.Items.Add(Brushes.Red)
        Me.Items.Add(Brushes.RosyBrown)
        Me.Items.Add(Brushes.RoyalBlue)
        Me.Items.Add(Brushes.SaddleBrown)
        Me.Items.Add(Brushes.Salmon)
        Me.Items.Add(Brushes.SandyBrown)
        Me.Items.Add(Brushes.SeaGreen)
        Me.Items.Add(Brushes.SeaShell)
        Me.Items.Add(Brushes.Sienna)
        Me.Items.Add(Brushes.Silver)
        Me.Items.Add(Brushes.SkyBlue)
        Me.Items.Add(Brushes.SlateBlue)
        Me.Items.Add(Brushes.SlateGray)
        Me.Items.Add(Brushes.Snow)
        Me.Items.Add(Brushes.SpringGreen)
        Me.Items.Add(Brushes.SteelBlue)
        Me.Items.Add(Brushes.Tan)
        Me.Items.Add(Brushes.Teal)
        Me.Items.Add(Brushes.Thistle)
        Me.Items.Add(Brushes.Tomato)
        Me.Items.Add(Brushes.Transparent)
        Me.Items.Add(Brushes.Turquoise)
        Me.Items.Add(Brushes.Violet)
        Me.Items.Add(Brushes.Wheat)
        Me.Items.Add(Brushes.White)
        Me.Items.Add(Brushes.WhiteSmoke)
        Me.Items.Add(Brushes.Yellow)
        Me.Items.Add(Brushes.YellowGreen)

    End Sub

    'Protected Overrides Sub RefreshItem(ByVal index As Integer)

    'End Sub

    Private Sub DrawItems(ByVal sender As Object, ByVal die As DrawItemEventArgs)
        ' If we have no index to draw, return.
        If die.Index = -1 Then
            Exit Sub
        End If

        ' The "sender" will be a combobox instance...if none, return.
        If sender Is Nothing Then
            Exit Sub
        End If

        ' Pull the controlling combobox object and prepare to render...
        Dim cmb As ComboBox = CType(sender, ComboBox)
        Dim selectedBrush As SolidBrush = CType(cmb.Items(die.Index), SolidBrush)
        Dim g As Graphics = die.Graphics

        ' If the item is selected, this draws the correct background color
        die.DrawBackground()
        die.DrawFocusRectangle()
        Dim rectPreviewBox As Rectangle = die.Bounds

        ' Draw the color's preview box
        rectPreviewBox.Offset(2, 2)
        rectPreviewBox.Width = 20
        rectPreviewBox.Height -= 4
        g.DrawRectangle(New Pen(die.ForeColor), rectPreviewBox)

        ' Get the appropriate Brush object for the selected color
        ' and fill the preview box.
        rectPreviewBox.Offset(1, 1)
        rectPreviewBox.Width -= 2
        rectPreviewBox.Height -= 2
        g.FillRectangle(selectedBrush, rectPreviewBox)

        ' Draw the name of the color selected. Note we'll clean up the returned
        ' type string. Normally it would come back:
        '
        ' "Color [Black]"
        '
        ' We'll clean it up to produce:
        '
        ' "Black"
        g.DrawString(selectedBrush.Color.ToString().Substring(7, selectedBrush.Color.ToString().Length - 8), Font, New SolidBrush(die.ForeColor), die.Bounds.X + 30, die.Bounds.Y + 1)
    End Sub

End Class
